text/0000-io-net-2.1.md

```diff
@@ -0,0 +1,527 @@
+    pub fn value(&self) -> c_int;
+}
+impl From<c_int> for Domain { ... }
+
+impl Type {
+    pub fn stream() -> Type;
+    pub fn datagram() -> Type;
+    pub fn raw() -> Type;
+    pub fn value(&self) -> c_int;
+}
+impl From<c_int> for Type { ... }
+
+impl From<c_int> for Protocol { ... }
+```
+
+The upside of this approach is that it is maximally flexible in terms of usage.
+All interactions with C APIs can be directly translated to Rust itself with ease
+and there is a clear layer of abstraction for what's going on here. The
+downsides of this approach, however, are:
+
+* It's unclear why the `TcpStream` and other refined primitives still exist.
+  This API is a union of the `TcpStream`, `TcpListener`, and `UdpSocket` APIs
+  and poses a new question of "should I use a Socket or TcpStream?".
+* The scope of this API is somewhat unclear, for example is `Socket` intended to
+  encompass Unix sockets as well? It attempts to via the `Into<RawSocketAddr>`
+  trait bound, but it's arguably not super ergonomic.
+* Many mis-usages of the API which would statically be prevented (such as
+  calling `accept` on a UDP socket) are allowed with a single `Socket` API.
+  Other examples of this are setting TCP options for a datagram socket.
+
+Overall the builder-based approach seemed to give the same level of flexibility
+while maintaining a clearly defined purpose of solely operating as builders for
+new sockets.
+
+### A note on typestate
+
+Note that there are many alternatives within this alternative itself (of having
+a unifying notion of a "socket"), such as using typestate to solve the mis-usage
+bullet point above. Sockets are often a classical use case of a state-transition
+diagram, so this can be a natural conclusion to come to.
+
+There are, however, not that many usages of typestate throughout the rest of the
+standard library today, and there are a number of ergonomic and API concerns
+which would be difficult to overcome.
+
+## Socket Options
+
+The detailed design section above outlined a convention for adding a pair of
+methods per socket option, but this has the drawback of leading to quite a large
+API surface area and it's tough to view the set of "options related methods" in
+a group. An alternative to the scheme proposed above would be something along
+the lines of the following:
+
+```rust
+mod net::opt {
+    pub trait Sockopt {
+        type Value;
+        fn get(socket: &Socket) -> io::Result<Self::Value>;
+        fn set(socket: &Socket, value: Self::Value) -> io::Result<()>;
+    }
+
+    // socket options
+    pub enum Keepalive {}       // SO_KEEPALIVE, Value = bool
+    pub enum ReuseAddress {}    // SO_REUSEADDR, Value = bool
+    pub enum Broadcast {}       // SO_BROADCAST, Value = bool
+    pub enum ReadTimeout {}     // SO_RCVTIMEO,  Value = Option<Duration>
+    pub enum WriteTimeout {}    // SO_SNDTIMEO,  Value = Option<Duration>
+
+    // tcp options
+    pub enum TcpNodelay {}      // TCP_NODELAY, Value = bool
```

Comment:
Certainly! This section, however, is just dedicated to outlining an alternative, but the standard platform-specific functionality we have in the rest of the standard library would allow binding this option from a `std::os::linux::net` extension trait.